siva/metadata.go

```diff
@@ -0,0 +1,156 @@
+package siva
+
+import (
+	"io/ioutil"
+	"os"
+
+	"github.com/ghodss/yaml"
+	billy "gopkg.in/src-d/go-billy.v4"
+)
+
+const (
+	// LibraryMetadataFile is the name of the file that holds library metadata.
+	LibraryMetadataFile = "library.yaml"
+)
+
+// LibraryMetadata holds information about the library.
+type LibraryMetadata struct {
+	Version int `json:"version"`
+}
+
+// ParseLibraryMetadata parses the yaml representation of library metadata.
+func ParseLibraryMetadata(d []byte) (*LibraryMetadata, error) {
+	var m LibraryMetadata
+
+	err := yaml.Unmarshal(d, &m)
+	if err != nil {
+		return nil, err
+	}
+
+	return &m, nil
+}
+
+// LoadLibraryMetadata reads and parses a library metadata file.
+func LoadLibraryMetadata(fs billy.Filesystem) (*LibraryMetadata, error) {
+	mf, err := fs.Open(LibraryMetadataFile)
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	defer mf.Close()
+
+	data, err := ioutil.ReadAll(mf)
+	if err != nil {
+		return nil, err
+	}
+
+	return ParseLibraryMetadata(data)
+}
+
+// Version describes a bookmark in the siva file.
+type Version struct {
+	// Offset is a position in the siva file.
+	Offset uint64 `json:"offset"`
+	// Size is block size of the version.
+	Size uint64 `json:"size,omiempty"`
+}
+
+// Versions holds a numbered list of bookmarks in the siva file.
+type Versions map[int]Version
+
+// LocationMetadata holds extra data associated with a siva file.
+type LocationMetadata struct {
+	Versions Versions `json:"versions"`
+}
+
+// ParseLocationMetadata parses the yaml representation of location metadata.
+func ParseLocationMetadata(d []byte) (*LocationMetadata, error) {
+	var m LocationMetadata
+
+	err := yaml.Unmarshal(d, &m)
+	if err != nil {
+		return nil, err
+	}
+
+	return &m, nil
+}
+
+// LoadLocationMetadata reads and parses a location metadata file.
+func LoadLocationMetadata(fs billy.Filesystem, path string) (*LocationMetadata, error) {
+	mf, err := fs.Open(path)
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	defer mf.Close()
+
+	data, err := ioutil.ReadAll(mf)
+	if err != nil {
+		return nil, err
+	}
+
+	return ParseLocationMetadata(data)
+}
+
+// LocationMetadataPath returns the path for a location metadata file.
+func LocationMetadataPath(path string) string {
```

Comment:
Not sure if it's worthy export this, kind of intern details on how we store the metadata

Reply:
It will be needed by the tool that updates sivas and updates the metadata.

Reply:
After a second thought you are right. Marshaling and Unmarshaling should be handled by the library and not an external tool.